File: Strings/longest_common_prefix.cpp
/*
Write a function to find the longest common prefix string amongst an array of strings.

If there is no common prefix, return an empty string "".

 

Example 1:

Input: strs = ["flower","flow","flight"]
Output: "fl"
Example 2:

Input: strs = ["dog","racecar","car"]
Output: ""
Explanation: There is no common prefix among the input strings.
 

Constraints:

1 <= strs.length <= 200
0 <= strs[i].length <= 200
strs[i] consists of only lowercase English letters.

Time complexity : O(strs)
*/

#include<bits/stdc++.h>

class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if(strs.size() == 0) return "";
        string ans = "";
        //  fix one string and check the common prefix of this string with other strings
        // s is the smallest string, so longest cant be greater than smallest string in array
        string s = *min_element(strs.begin(), strs.end());
        for(int i = 0; i < s.size(); i++){
            for(int j = 0; j < strs.size(); j++){
                if(s[i] != strs[j][i])
                    return ans;
            }
            ans.push_back(s[i]);
        }
        return ans;
    }
};

File: Strings/LongestCommonPrefix.java
/**
 * Given the array of strings A, you need to find the longest string S, which is the prefix of ALL the strings in the array.
 *
 * The longest common prefix for a pair of strings S1 and S2 is the longest string S which is the prefix of both S1 and S2.
 *
 * Example: the longest common prefix of "abcdefgh" and "abcefgh" is "abc".
 *
 *
 *
 * Problem Constraints
 * 0 <= sum of length of all strings <= 1000000
 *
 *
 *
 * Input Format
 * The only argument given is an array of strings A.
 *
 *
 *
 * Output Format
 * Return the longest common prefix of all strings in A.
 *
 *
 *
 * Example Input
 * Input 1:
 *
 * A = ["abcdefgh", "aefghijk", "abcefgh"]
 * Input 2:
 *
 * A = ["abab", "ab", "abcd"];
 *
 *
 * Example Output
 * Output 1:
 *
 * "a"
 * Output 2:
 *
 * "ab"
 *
 *
 * Example Explanation
 * Explanation 1:
 *
 * Longest common prefix of all the strings is "a".
 * Explanation 2:
 *
 * Longest common prefix of all the strings is "ab".
 */
package Strings;

public class LongestCommonPrefix {
    public static void main(String[] args) {
        String[] strings = {"ab", "abab", "ab", "abcd"};

        String res = solve(strings);
        System.out.println(res);
    }
    public static String solve(String[] strings) {
        if (strings == null || strings.length == 0) return "";

        for (int i = 0; i < strings[0].length(); i++) {
            char currentChar = strings[0].charAt(i);

            for (int j = 1; j < strings.length; j++) {
                if (i == strings[j].length() || strings[j].charAt(i) != currentChar)
                        return strings[0].substring(0, i);
            }
        }
        return strings[0];
    }
}
